lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
+    Path(Path), OpenEC(), OS(Path, OpenEC), OutputtingArguments(false) {
+    revng_assert(!OpenEC);
+    printHeader();
+  }
+
+  ~PipelineCTracer() {
+    OS.flush();
+    OS.close();
+  }
+
+private:
+  void printHeader() {
+    OS << "version: 1\n";
+    OS << "commands:\n";
+    OS.flush();
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
+  }
+
+public:
+  void functionPrelude(const llvm::StringRef Name) {
+    OS << "- name: " << Name << "\n";
+    OS << "  arguments:\n";
+    OutputtingArguments = true;
+    OS.flush();
+  }
+
+  void newArgument() {
+    OS << "  - ";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t> ||
+           std::is_same_v<T, uint32_t> ||
+           std::is_same_v<T, uint64_t>
+  // clang-format on
+  void printValue(const T &Int) {
+    OS << Int << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, char *> ||
+           std::is_same_v<T, const char *>
+  // clang-format on
+  void printValue(T String) {
+    if (OutputtingArguments)
+      OS << reprString(String);
+    else
+      OS << "P" << (void *) String;
+    OS << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  requires std::is_same_v<T, bool>
```

Comment:
Can't we?
```suggestion
```

Reply:
That would allow objects that are castable to bool to select this overload, which should not happen

Reply:
Right, but this is true for anywhere we have a `bool` argument, but we don't turn it each `bool` into a template to avoid this pitfall. This would be more like a compiler warning ([this](https://releases.llvm.org/13.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/readability-implicit-bool-conversion.html)?).

Or is this case somehow special? If not, let's just have `bool`.

Reply:
Sure, but in this case every argument is called via `printValue`, so I would like to be sure the overload selector does the right thing.
Say, even if a bit contrived, that someone introduces a `char16_t` argument, and they forget to implement/enrich a `printValue` function. The compilation finishes because the argument was cast to `bool` and no one notices until it's too late

Reply:
Added